prow/cmd/deck/job_history.go

```diff
@@ -0,0 +1,274 @@
+	path := path.Join(root, latestBuildFile)
+	data, err := readObject(bkt.Object(path))
+	if err != nil {
+		return -1, fmt.Errorf("failed to read latest build number: %v", err)
+	}
+	n, err := strconv.Atoi(string(data))
+	if err != nil {
+		return -1, fmt.Errorf("failed to parse latest build number: %v", err)
+	}
+	return n, nil
+}
+
+// resolve sym links into the actual log directory for a particular test run
+func resolveSymLink(bkt *storage.BucketHandle, symLink string) (string, error) {
+	data, err := readObject(bkt.Object(symLink))
+	if err != nil {
+		return "", fmt.Errorf("failed to read %s: %v", symLink, err)
+	}
+	// strip gs://<bucket-name> from global address `u`
+	u := string(data)
+	prefixRe := regexp.MustCompile("gs://.*?/")
+	return prefixRe.ReplaceAllString(u, ""), nil
+}
+
+func getDir(bkt *storage.BucketHandle, root, id string) (string, error) {
+	bAttrs, err := bkt.Attrs(context.Background())
+	if err != nil {
+		return "", fmt.Errorf("failed to get bucket name: %v", err)
+	}
+	bktName := bAttrs.Name
+	if strings.HasPrefix(root, logsPrefix) {
+		return path.Join(bktName, root, id), nil
+	}
+	symLink := path.Join(root, id+".txt")
+	dir, err := resolveSymLink(bkt, symLink)
+	if err != nil {
+		return "", fmt.Errorf("failed to resolve sym link: %v", err)
+	}
+	return path.Join(bktName, dir), nil
+}
+
+func getObject(bkt *storage.BucketHandle, root, id, fname string) *storage.ObjectHandle {
+	if strings.HasPrefix(root, logsPrefix) {
+		p := path.Join(root, id, fname)
+		return bkt.Object(p)
+	}
+	symLink := path.Join(root, id+".txt")
+	dir, err := resolveSymLink(bkt, symLink)
+	if err != nil {
+		return &storage.ObjectHandle{}
+	}
+	return bkt.Object(path.Join(dir, fname))
+}
+
+func fileExists(bkt *storage.BucketHandle, root, id, fname string) bool {
+	obj := getObject(bkt, root, id, fname)
+	_, err := obj.Attrs(context.Background())
+	return err == nil
+}
+
+func readStarted(bkt *storage.BucketHandle, root, id string) (jobs.Started, error) {
+	s := jobs.Started{}
+	sobj := getObject(bkt, root, id, "started.json")
+	sdata, err := readObject(sobj)
+	if err != nil {
+		return s, fmt.Errorf("failed to read started.json for build %s: %v", id, err)
+	}
+	err = json.Unmarshal(sdata, &s)
+	if err != nil {
+		return s, fmt.Errorf("failed to parse started.json for build %s: %v", id, err)
+	}
+	return s, nil
+}
+
+func readFinished(bkt *storage.BucketHandle, root, id string) (jobs.Finished, error) {
+	f := jobs.Finished{}
+	fobj := getObject(bkt, root, id, "finished.json")
+	fdata, err := readObject(fobj)
+	if err != nil {
+		return f, fmt.Errorf("failed to read finished.json for build %s: %v", id, err)
+	}
+	err = json.Unmarshal(fdata, &f)
+	if err != nil {
+		return f, fmt.Errorf("failed to parse finished.json for build %s: %v", id, err)
+	}
+	return f, nil
+}
+
+// Gets job history from the GCS bucket specified in config.
+func getJobHistory(url *url.URL, config *config.Config, gcsClient *storage.Client) (JobHistoryTemplate, error) {
```

Comment:
nit: This could use a few more comments and might benefit from moving some of the logic into its helper functions for clarity and easier unit testing.